cpp/src/arrow/tensor/coo_converter.cc

```diff
@@ -180,34 +181,29 @@ class SparseCOOTensorConverter : private SparseTensorConverterMixin {
 
     ARROW_ASSIGN_OR_RAISE(auto indices_buffer,
                           AllocateBuffer(index_elsize * ndim * nonzero_count, pool_));
-    uint8_t* indices = indices_buffer->mutable_data();
 
     ARROW_ASSIGN_OR_RAISE(auto values_buffer,
                           AllocateBuffer(value_elsize * nonzero_count, pool_));
-    uint8_t* values = values_buffer->mutable_data();
 
-    const uint8_t* tensor_data = tensor_.raw_data();
+    auto* values = values_buffer->mutable_data_as<ValueCType>();
+    const auto* tensor_data = tensor_.data()->data_as<ValueCType>();
+    auto* indices = indices_buffer->mutable_data_as<IndexCType>();
     if (ndim <= 1) {
       const int64_t count = ndim == 0 ? 1 : tensor_.shape()[0];
       for (int64_t i = 0; i < count; ++i) {
-        if (std::any_of(tensor_data, tensor_data + value_elsize, IsNonZero)) {
-          AssignIndex(indices, i, index_elsize);
-          std::copy_n(tensor_data, value_elsize, values);
-
-          indices += index_elsize;
-          values += value_elsize;
+        if (is_not_zero<ValueType>(*tensor_data)) {
+          *indices++ = static_cast<IndexCType>(i);
+          *values++ = *tensor_data;
         }
-        tensor_data += value_elsize;
+        ++tensor_data;
       }
     } else if (tensor_.is_row_major()) {
-      DISPATCH(CONVERT_ROW_MAJOR_TENSOR, index_elsize, value_elsize, indices, values,
-               nonzero_count);
+      ConvertRowMajorTensor<IndexType, ValueType>(tensor_, indices, values);
     } else if (tensor_.is_column_major()) {
-      DISPATCH(CONVERT_COLUMN_MAJOR_TENSOR, index_elsize, value_elsize, indices, values,
-               nonzero_count);
+      ConvertColumnMajorTensor<IndexType, ValueType>(tensor_, indices, values,
+                                                     nonzero_count);
     } else {
-      DISPATCH(CONVERT_STRIDED_TENSOR, index_elsize, value_elsize, indices, values,
-               nonzero_count);
+      ConvertStridedTensor<IndexType, ValueType>(tensor_, indices, values);
```

Comment:
Switching from DISPATCH to templates seems ok to me, but perhaps @pitrou can double-check?